### PR TITLE
Bounds - Preparations Update

### PR DESCRIPTION
**Pull request by Fusezion**
This pr is the start of #400 which includes the start to bounds enhancements. I have listed what I've done in this PR below along with what I want to finish adding before we go ahead and merge this.

- [x] ExitBoundEvent and EnterBoundEvent renamed to BoundExitEvent and BoundEnterEvent
- [x] BoundEvent super class added and above events extending it
- [x] Update to BoundEvents classes' event values 
- [x] Enhancements to the `ExprBoundAllBounds` and `ExprBoundAtLocation` classes
- [x] Config settings to change which events are checked by `BoundBorderListener`
- [x] Even more events added to `BoundBorderListener`
- [x] plurals in some bound syntax
- [x] New expression, last created bound
- [x] Clone/Copy bound
- [x] Fix `delete bounds` breaking change, if possible
- [x] Fix ExprLastCreatedBound, retaining bound values of deleted bounds

Related Issue(s): #353

### CODE REVIEW

**Comment by Fusezion** [comment]
With only bound copy section being needed, I'm going ahead and marking this ready for review. Should hopefully finish this tonight if not tomorrow.

File: src/main/java/com/shanebeestudios/skbee/elements/bound/expressions/ExprAllBounds.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+    @Override
+    public @NotNull String toString(@Nullable Event event, boolean debug) {
+        String ID = this.ID ? " ids" : "s";
+        return switch (boundType) {
+            case TRUE -> "all non-temporary bound" + ID;
+            case FALSE -> "all temporary bound" + ID;
+            case UNKNOWN -> "all bound" + ID;
+        };
+    }
```
Comment (ShaneBeee) [inline src/main/java/com/shanebeestudios/skbee/elements/bound/expressions/ExprAllBounds.java:87]: simplify?!?!
```suggestion
    public @NotNull String toString(@Nullable Event event, boolean debug) {
        String ID = this.ID ? " ids" : "s";
        return switch (boundType) {
            case TRUE -> "all non-temporary bound";
            case FALSE -> "all temporary bound";
            case UNKNOWN -> "all bound";
        }  + ID;
    }
```

File: src/main/java/com/shanebeestudios/skbee/elements/bound/expressions/ExprBoundsAtLocation.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    @Override
+    public @NotNull String toString(Event event, boolean debug) {
+        String ID = this.ID ? " ids" : "s";
+        String atLocation = " at " + locations.toString(event, debug);
+        return switch (boundType) {
+            case TRUE -> "all non-temporary bound" + ID + atLocation;
+            case FALSE -> "all temporary bound" + ID + atLocation;
+            case UNKNOWN -> "all bound" + ID + atLocation;
+        };
+    }
```
Comment (ShaneBeee) [inline src/main/java/com/shanebeestudios/skbee/elements/bound/expressions/ExprBoundsAtLocation.java:93]: simplify again?!?!
```suggestion
    public @NotNull String toString(Event event, boolean debug) {
        String ID = this.ID ? " ids" : "s";
        String atLocation = " at " + locations.toString(event, debug);
        return switch (boundType) {
            case TRUE -> "all non-temporary bound";
            case FALSE -> "all temporary bound";
            case UNKNOWN -> "all bound";
        } + ID + atLocation;
    }
```

File: src/main/java/com/shanebeestudios/skbee/elements/bound/expressions/ExprBoundsInWorld.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    @Override
+    public String toString(@Nullable Event event, boolean debug) {
+        String ID = this.ID ? " ids" : "s";
+        String inWorld = " in world " + worlds.toString(event, debug);
+        return switch (boundType) {
+            case TRUE -> "all non-temporary bound" + ID + inWorld;
+            case FALSE -> "all temporary bound" + ID + inWorld;
+            case UNKNOWN -> "all bound" + ID + inWorld;
+        };
+    }
```
Comment (ShaneBeee) [inline src/main/java/com/shanebeestudios/skbee/elements/bound/expressions/ExprBoundsInWorld.java:92]: one more time
```suggestion
    public String toString(@Nullable Event event, boolean debug) {
        String ID = this.ID ? " ids" : "s";
        String inWorld = " in world " + worlds.toString(event, debug);
        return switch (boundType) {
            case TRUE -> "all non-temporary bound";
            case FALSE -> "all temporary bound";
            case UNKNOWN -> "all bound";
        } + ID + inWorld;
    }
```

File: src/main/java/com/shanebeestudios/skbee/elements/bound/expressions/ExprLastCreatedBound.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    @Override
+    public Class<? extends Bound> getReturnType() {
+        return Bound.class;
+    }
+
+    @Override
+    public String toString(@Nullable Event event, boolean b) {
+        return "lastly created bound";
+    }
+}
```
Comment (ShaneBeee) [inline src/main/java/com/shanebeestudios/skbee/elements/bound/expressions/ExprLastCreatedBound.java:65]:
```suggestion
    }

}
```

File: src/main/java/com/shanebeestudios/skbee/elements/bound/sections/EffSecBoundCopy.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+import com.shanebeestudios.skbee.config.BoundConfig;
+import com.shanebeestudios.skbee.elements.bound.expressions.ExprLastCreatedBound;
+import org.bukkit.event.Event;
+import org.eclipse.jdt.annotation.Nullable;
+
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+@Name("Bound - Copy Bound")
+@Description("Create an exact replica of an existing bound with a new id")
```
Comment (ShaneBeee) [inline src/main/java/com/shanebeestudios/skbee/elements/bound/sections/EffSecBoundCopy.java:28]:
```suggestion
@Description("Create an exact replica of an existing bound with a new id.")
```

**Review by ShaneBeee — CHANGES_REQUESTED** [review]
just a few small changes then I think its good to go

File: src/main/java/com/shanebeestudios/skbee/api/listener/BoundBorderListener.java [hunk 6]
```diff
@@ -68,19 +139,50 @@ private void onMount(EntityMountEvent event) {
...
+
+    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
     private void onVehicleEnter(VehicleEnterEvent event) {
+        if (!this.VEHICLE_ENTER) return;
         if (event.getEntered() instanceof Player player) {
             Location from = player.getLocation();
             Location to = event.getVehicle().getLocation();
-            if (preventBoundMovement(player, from, to)) {
-                event.setCancelled(true);
-            }
```
Comment (sovdeeth) [inline src/main/java/com/shanebeestudios/skbee/api/listener/BoundBorderListener.java:78]: Was this intended @Fusezion

Comment (sovdeeth) [reply src/main/java/com/shanebeestudios/skbee/api/listener/BoundBorderListener.java:78]: image
Seems to leave a useless listener

**Pull request merged**